packages/core/src/services/users/invite.ts

```diff
@@ -17,34 +23,65 @@ export async function inviteUser(
     author,
   }: {
     email: string
-    name: string
+    name?: string
     workspace: Workspace
     author: User
   },
   db = database,
-) {
+): Promise<TypedResult<InviteUserOutcome, Error>> {
   let user = await db.query.users.findFirst({ where: eq(users.email, email) })
 
-  return Transaction.call(async (tx) => {
-    if (!user) {
-      const result = await createUser({ email, name })
-      user = result.unwrap()
-    }
+  return Transaction.call<InviteUserOutcome>(async (tx) => {
+    if (env.INVITE_ONLY && !user) {
+      const invitationResult = await createInvitation(
+        {
+          email,
+          invitedByUserId: author.id,
+          workspaceId: workspace.id, // Pass as number
+        },
+        tx,
+      )
 
-    await createMembership({ author, user, workspace }, tx).then((r) =>
-      r.unwrap(),
-    )
+      if (!invitationResult.ok) {
+        return Result.error(invitationResult.error || new Error('Failed to create invitation.'))
+      }
 
-    publisher.publishLater({
-      type: 'userInvited',
-      data: {
-        invited: user,
-        invitee: author,
-        userEmail: author.email,
-        workspaceId: workspace.id,
-      },
-    })
+      const invitation = invitationResult.unwrap()
+      publisher.publishLater({
+        type: 'userInvitationCreated',
+        data: {
+          invitation: invitation,
+          invitedByUserEmail: author.email,
+          workspaceId: workspace.id,
+        },
+      })
+      return Result.ok({ status: 'invitation_created', invitation })
+    } else {
+      // Either INVITE_ONLY is false, or user already exists
+      if (!user) {
+        // INVITE_ONLY is false and user does not exist: Create user
+        const createUserResult = await createUser({ email, name: name ?? '' }, tx)
+        if (!createUserResult.ok) {
+          return Result.error(createUserResult.error || new Error('Failed to create user.'))
+        }
+        user = createUserResult.unwrap()
+      }
 
-    return Result.ok(user)
+      // User exists (either found or just created), add to workspace
+      const membershipResult = await createMembership({ author, user: user!, workspace }, tx)
+      if (!membershipResult.ok) {
+        return Result.error(membershipResult.error || new Error('Failed to create membership.'))
+      }
+      
+      // publisher.publishLater({ // Commenting out for now
+      //   type: 'userAddedToWorkspace',
+      //   data: {
+      //     addedUserId: user!.id,
+      //     invitingUserId: author.id,
+      //     workspaceId: workspace.id, // Ensure workspace.id is string
+      //   },
+      // })
+      return Result.ok({ status: 'user_added_to_workspace', user: user! })
+    }
```

Comment:
I know this service didn't had tests. But before was simple now it has a couple of branches. Please add tests